README.rst
**********************************************************
BridgeDB |Latest Version| |Build Status| |Coverage Status|
**********************************************************

BridgeDB is a collection of backend servers used to distribute `Tor Bridges
<https://www.torproject.org/docs/bridges>`__. Currently, it mainly consists of
a webserver with `an HTTPS interface <https://bridges.torproject.org>`__,
`an email responder <mailto:bridges@torproject.org>`__, and an SQLite database.

.. |Latest Version| image:: https://pypip.in/version/bridgedb/badge.svg?style=flat
   :target: https://pypi.python.org/pypi/bridgedb/
.. |Build Status| image:: https://travis-ci.org/sysrqbci/bridgedb.svg
   :target: https://travis-ci.org/sysrqbci/bridgedb
.. |Coverage Status| image:: https://coveralls.io/repos/github/sysrqbci/bridgedb/badge.svg?branch=develop
   :target: https://coveralls.io/github/sysrqbci/bridgedb?branch=develop



.. image:: doc/sphinx/source/_static/bay-bridge.jpg
   :scale: 80%
   :align: center


.. contents::
   :backlinks: entry


=====================
What are Tor Bridges?
=====================

`Tor Bridges <https://www.torproject.org/docs/bridges>`__ are special
Tor relays which are not listed in the public relay directory. They are
used to help circumvent `censorship <https://ooni.torproject.org>`__ by
providing users with connections to the public relays in the Tor
network.

Tor Bridges are different from normal relays in another important way:
they can run what are called *Pluggable* *Transports*.

-----------------------------
What's a Pluggable Transport?
-----------------------------

A `Pluggable
Transport <https://www.torproject.org/docs/pluggable-transports.html.en>`__
is a program which is *pluggable* — meaning that it is meant to work
with lots of other anonymity and censorship circumvention software, not
just Tor — and is a *transport* — meaning that it transports your
internet traffic, usually in a way which makes it look different. For
example,
`Obfsproxy <https://www.torproject.org/projects/obfsproxy.html.en>`__ is
a Pluggable Transport which disguises your traffic by adding an
obfuscating layer of encryption.

---------------------
So how do I use this?
---------------------

Well, probably, you don't. But if you're looking for bridges, you can
use `the web interface <https://bridges.torproject.org>`__ of the
BridgeDB instance deployed by the Tor Project, which has instructions on
getting the Pluggable Transports-capable Tor Browser Bundle, as well as
instructions for getting extra Bridges.


================
Maintainer Setup
================

If you'd like to hack on BridgeDB, you might wish to read BridgeDB's
`developer documentation <https://pythonhosted.org/bridgedb/>`__.  The rest of
this document mainly concerns mainenance and installation instructions.

-----------------------------
Dependencies and installation
-----------------------------

BridgeDB requires the following OS-level dependencies:

-  python>=2.7
-  python-dev
-  `python3-dkim <https://pypi.org/project/dkimpy/>`__ (it contains the ``dkimverify`` binary)
-  build-essential
-  gnupg (preferrably, gnupg2)
-  OpenSSL>=1.0.1g
-  `SQLite3 <http://www.maxmind.com/app/python>`__
-  `MaxMind GeoIP <https://www.maxmind.com/en/geolocation_landing>`__
-  libgeoip-dev
-  geoip-database
-  `python-setuptools <https://pypi.python.org/pypi/setuptools>`__
-  libjpeg-dev
-  `flog <https://packages.debian.org/jessie/flog>`__ (only required if bridgedb
   is invocated with the ``run-bridgedb`` `script
   <https://gitweb.torproject.org/project/bridges/bridgedb-admin.git/tree/bin/run-bridgedb>`__)

As well as any Python dependencies in the ``requirements.txt`` file.

.. note: There are additional dependencies for things like running the test
    suites, building BridgeDB's developer documentation, etc. Read on for more
    info if you wish to enable addition features.


------------------
Deploying BridgeDB
------------------

BridgeDB should work with or without a Python virtualenv.

-  Install Python 2.7, and other OS-level dependencies. On Debian, you
   can do::

         sudo apt-get install build-essential openssl python python-dev \
           python-setuptools sqlite3 gnupg2 libgeoip-dev geoip-database


-  Install Pip 1.3.1 or later. Debian has this version, but if for some
   reason that or a newer version isn't available, the easiest way to
   install a newer Pip is to use the Pip development teams's `getpip
   script <https://raw.github.com/pypa/pip/master/contrib/get-pip.py>`__::

         wget https://raw.github.com/pypa/pip/master/contrib/get-pip.py
         sudo python get-pip.py


-  **(virtualenv installs only)** Use Pip to install virtualenv and
   `virtualenvwrapper <https://virtualenvwrapper.readthedocs.org>`__::

         sudo pip install --upgrade virtualenv virtualenvwrapper


-  **(virtualenv installs only)** Configure virtualenvwrapper and create a
   virtualenv for BridgeDB::

         WORKON_HOME=${HOME}/.virtualenvs
         export WORKON_HOME
         mkdir -p $WORKON_HOME
         source $(which virtualenvwrapper.sh)
         git clone https://git.torproject.org/bridgedb.git && cd bridgedb
         mkvirtualenv -a $PWD -r requirements.txt --unzip-setuptools --setuptools bridgedb

   From now on, to use BridgeDB's virtualenv, just do ``$ workon bridgedb``
   (after sourcing virtualenvwrapper.sh, as before). To exit the virtualenv
   without exiting the shell, do ``$ deactivate``.


-  **(virtualenv installs only)** To install, set PYTHONPATH to include the
   root directory of the virtualenv::

         export PYTHONPATH=$PYTHONPATH:${VIRTUAL_ENV}/lib/python2.7/site-packages


-  Then, proceed as usual::

         python setup.py install --record installed-files.txt


============================
Enabling additional features
============================

------------
Translations
------------

For general information on the translation process, take a look at
`our translation guidelines for developers
<https://trac.torproject.org/projects/tor/wiki/doc/translation/developers>`__.

**Using New Translations**:

This should be done when newly completed translations are available in
Transifex.

Piece of cake. Running ``maint/get-completed-translations`` will take
care of cloning *only* the ``bridgedb_completed`` branch of Tor's
`translations repo <https://gitweb.torproject.org/translation.git>`__
and placing all the updated files in their correct locations.

-------

**Requesting Translations for Altered/Added Source Code**:

This should be done whenever any of the strings requiring translation --
``_("the ones inside the weird underscore function, like this")`` -- are
changed, or new ones are added. See ``lib/bridgedb/strings.py``.

Translations for Tor Project repos are kept `in a separate repo
<https://gitweb.torproject.org/translation.git>`__. To request new or updated
translations, you'll need to extract the strings from BridgeDB's source code
into our ./i18n/templates/bridgedb.pot template, and then commit it to our
``develop`` branch.  From there, the translation system takes over and
eventually, new translations will be available.  To extract all strings from
BridgeDB's source, run::

         python setup.py extract_messages

Transifex uses the resulting file ./i18n/templates/bridgedb.pot (and this file
only) as input and fetches it from BridgeDB's ``develop`` branch, so we don't
need to release a new BridgeDB version (which we only do in the ``master``
branch) to request translations.

-------

--------------
Enabling HTTPS
--------------

Create a self-signed certificate with::

         scripts/make-ssl-cert

Or, place an existing certificate in the path specified in bridgedb.conf
by the ``HTTPS_CERT_FILE`` option, and a private key where
``HTTPS_KEY_FILE`` points to. The defaults are 'cert' and 'privkey.pem',
respectively.


------------------------
Enabling CAPTCHA Support
------------------------

BridgeDB has two ways to use CAPTCHAs on webpages. The first uses reCaptcha_,
an external Google service (this requires an account with them), which
BridgeDB fetches the CAPTCHAs images from for each incoming request from a
client. The second method uses a local cache of pre-made CAPTCHAs, created by
scripting Gimp using gimp-captcha_. The latter cannot easily be run on
headless server, unfortunately, because Gimp requires an X server to be
installed.

.. _reCaptcha: https://www.google.com/recaptcha
.. _gimp-captcha: https://github.com/isislovecruft/gimp-captcha


**reCaptcha**

To enable fetching CAPTCHAs from the reCaptcha API server, set these
options in bridgedb.conf::

      RECAPTCHA_ENABLED
      RECAPTCHA_PUB_KEY
      RECAPTCHA_SEC_KEY

-------

**gimp-captcha**

To enable using a local cache of CAPTCHAs, set the following options::

      GIMP_CAPTCHA_ENABLED
      GIMP_CAPTCHA_DIR
      GIMP_CAPTCHA_HMAC_KEYFILE
      GIMP_CAPTCHA_RSA_KEYFILE

-------

--------------------
GnuPG email signing:
--------------------

In your ``bridgedb.conf`` file, make sure that::

      EMAIL_GPG_SIGNING_ENABLED = True

and edit the following option to add the full fingerprint of the GnuPG key
that should be used to by BridgeDB to sign outgoing emails::

      EMAIL_GPG_PRIMARY_KEY_FINGERPRINT

The key specified by ``EMAIL_GPG_PRIMARY_KEY_FINGERPRINT`` can be a master
key, or a subkey (with or without the private portions of its corresponding
master key), but it **must** be inside the ``secring.gpg`` and ``pubring.gpg``
keyrings inside the directory specified in the ``bridgedb.conf`` option::

      EMAIL_GPG_HOMEDIR

If the key has requires a passphrase for signing, you'll also need to set
either of::

      EMAIL_GPG_PASSPHRASE
      EMAIL_GPG_PASSPHRASE_FILE


----------------------------------------------------------
Preventing already-blocked bridges from being distributed:
----------------------------------------------------------

Uncomment or add ``COUNTRY_BLOCK_FILE`` to your bridgedb.conf. This file
should contain one bridge entry per line, in the format::

      fingerprint <bridge fingerprint> country-code <country code>

If the ``COUNTRY_BLOCK_FILE`` file is present, bridgedb will filter
blocked bridges from the responses it gives to clients requesting
bridges.


================
Testing BridgeDB
================

Before running to any of BridgeDB's test suites, make sure you have the
additional dependencies in the Pip requirements file,
``.test.requirements.txt`` installed::

      pip install -r .test.requirements.txt

To create a bunch of fake bridge descriptors to test BridgeDB, do::

      bridgedb mock [-n NUMBER_OF_DESCRIPTORS]

To run the test suites, do::

      make coverage

If you just want to run the tests, and don't care about code coverage
statistics, see the ``bridgedb trial`` and ``bridgedb test`` commands.


================
Running BridgeDB
================

To run BridgeDB, simply make any necessary changes to bridgedb.conf, and do::

      bridgedb

And remember that all files/directories in ``bridgedb.conf`` are assumed
relative to the runtime directory. By default, BridgeDB uses the current
working directory; you can, however specify an a different runtime
directory::

      bridgedb -r /srv/bridges.torproject.org/run

Make sure that the files and directories referred to in bridgedb.conf
exist. However, many of them, if not found, will be touched on disk so
that attempts to read/write from/to them will not raise excessive
errors.


----------------------------------------------
Reloading Bridges From Their Descriptor Files:
----------------------------------------------

When you have new lists of bridges from the Bridge Authority, replace
the old files and do::

      bridgedb --reload

Or just give it a SIGHUP::

      kill -s SIGHUP `cat .../run/bridgedb.pid`


=========================
Using a BridgeDB Instance
=========================

Obviously, you'll have to feed it bridge descriptor files from a
BridgeAuthority. There's currently only one BridgeAuthority in the entire
world, but Tor Project is, of course, very interested in adding support for
multiple BridgeAuthorities so that we can scale our own network, and make it
easier for individual and organisations who wish to run a lot of Tor bridge
relays have an easier time distributing those bridges themselves (if they wish
to do so). If you'd like to fund our work on this, please contact
tor-dev@lists.torproject.org!

----------------------------------
Accessing the HTTPS User Interface
----------------------------------

Just connect to the appropriate port. (See the ``HTTPS_PORT`` and
``HTTP_UNENCRYPTED_PORT`` options in the ``bridgedb.conf`` file.)

The HTTPS interface for our BridgeDB instance can be found `here
<https://bridges.torproject.org>`__.


----------------------------------
Accessing the Email User Interface
----------------------------------

Any mail sent to the ``EMAIL_PORT`` with a destination username as defined by
the ``EMAIL_USERNAME`` configuration option (the default is ``'bridge'``, e.g.
bridges@...) and sent from an ``@riseup.net`` or ``@gmail.com`` address (by
default, but configurable with the ``EMAIL_DOMAINS`` option).

You can email our BridgeDB instance `here <mailto:bridges@torproject.org>`__.


----------------------------
Accessing the Moat Interface
----------------------------

Moat is a bridge distributor for requesting bridges through `Tor Launcher's
<https://gitweb.torproject.org/tor-launcher.git/>`__ user interface.

The following describes the Moat API, version 0.1.0.

The client and server both MUST conform to `JSON-API <http://jsonapi.org/>`__.

The client SHOULD direct all requests via the Meek reflector at ``MEEK_REFECTOR``.
..
   XXX meek reflector URL

Requesting Bridges
""""""""""""""""""

The client MUST send a ``POST /moat/fetch`` containing the following JSON::

    {
      "data": [{
        "version": "0.1.0",
        "type": "client-transports",
        "supported": [ "TRANSPORT", "TRANSPORT", ... ],
      }]
    }

where:

* ``TRANSPORT`` is a string identifying a transport, e.g. ``"obfs3"`` or
  ``"obfs4"``.  Currently supported transport identifiers are:
  - ``"vanilla"``
  - ``"fte"``
  - ``"obfs3"``
  - ``"obfs4"``
  - ``"scramblesuit"``



CHANGELOG
        * FIXES https://bugs.torproject.org/33008
        Add an info page, available at bridges.torproject.org/info.  Relay
        Search links to this info page to explain to bridge operators what their
        bridge distribution mechanism means.

Changes in version 0.9.4 - 2020-02-19

        * FIXES https://bugs.torproject.org/30946
        This patch ports BridgeDB to Python 3.  Python 2 is no longer supported
        since Jan 1, 2020.

Changes in version 0.9.3 - 2020-02-18

        * FIXES <https://bugs.torproject.org/33299>
        This patch disables the distribution of FTE, ScrambleSuit, and obfs3.

Changes in version 0.9.2 - 2020-02-04

        * FIXES <https://bugs.torproject.org/31427>
        This patch updates the AUTHORS file, HACKING.md, contact information,
        the Trac URL to report bugs, our instructions on BridgeDB's landing
        page, and it fixes a small bug in descriptor generation.

Changes in version 0.9.1 - 2019-10-29

        * FIXES https://bugs.torproject.org/32203
        The metrics code used to weed out vanilla bridges, so they did not show
        up in our metrics.  This patch fixes this issue.

        * FIXES https://bugs.torproject.org/32134
        While implementing our language switcher (#26543), we added a new
        string, "Language", that requires translations.  This patch adds a new
        translation request and also updates our instructions on how to request
        new translations.

        * FIXES https://bugs.torproject.org/32105
        Mention an undocumented OS-level dependency: python3-dkim.

Changes in version 0.9.0 - 2019-10-16

        * FIXES https://bugs.torproject.org/26543
        Implement a language switcher that allows users to override the locale
        that BridgeDB automatically selects by inspecting the client's request
        headers.

Changes in version 0.8.3 - 2019-10-03

        * FIXES https://bugs.torproject.org/31903
        Update existing translations and request new translations.  Thanks to
        all volunteers who helped translate BridgeDB!

        * FIXES https://bugs.torproject.org/31780
        We implemented BridgeDB's metrics in #9316 but haven't specified its
        format until now.  In addition to adding a specification, this patch
        also makes our implementation consistent with our (slightly updated)
        specification.

        * FIXES https://bugs.torproject.org/29484
        Update BridgeDB's requirements to the latest respective versions.  Among
        others, this patch set updates Twisted to 19.7.0, pyOpenSSL to 19.0.0,
        and replaces (the abandoned) PyCrypto with PyCryptodome, which fixes
        security vulnerabilities.

Changes in version 0.8.2 - 2019-09-20

        Updated translations for the following languages:
        bn, da, eo, fa, it, ko, nl, pt_BR, pt_PT, sr, zh_CN.

Changes in version 0.8.1 - 2019-09-11

        * FIXES https://bugs.torproject.org/17626
        BridgeDB gets confused when users reply to a "get help" email.  The
        issue is that BridgeDB interprets commands anywhere in the email body,
        even if it's in quoted text.  To fix this issue, we are ignoring
        commands whose email body line starts with a '>' character, which is
        typically used for email quotes.

        * FIXES https://bugs.torproject.org/28533
        The frontdesk is seeing plenty of empty bogus emails.  This fix removes
        the email links and instead encourages users to take a look at the Tor
        Browser Manual and at our Support Portal.

Changes in version 0.8.0 - 2019-08-20

        * FIXES https://bugs.torproject.org/9316
        Make BridgeDB export usage metrics every 24 hours.  At the end of each
        24-hour measurement interval, BridgeDB will append usage metrics to the
        file METRICS_FILE, which is configured in bridgedb.conf.  Our metrics
        keep track of the number of (un)successful requests per transport type
        per country code (or email provider) per distribution method.  This way,
        we get to learn that, say, over the last 24 hours there were 31-40 users
        in Iran who successfully requested an obfs4 bridge over Moat.

        * FIXES #26542 https://bugs.torproject.org/26542
        Make BridgeDB distribute vanilla IPv6 bridges again.

        * FIXES #22755 https://bugs.torproject.org/22755
        Use stem instead of leekspin to create test descriptors.  We now don't
        need to depend on leekspin anymore.

        * FIXES #31252 https://bugs.torproject.org/31252
        Add an anti-bot mechanism that allows us to detect bots by matching
        HTTP request headers for blacklisted patterns.  For example, bots may
        have their Accept-Language set to "Klingon".  Blacklisted patterns are
        configured in BLACKLISTED_REQUEST_HEADERS_FILE.  When BridgeDB detects
        a bot request, we can answer their request with a decoy bridge that's
        only handed out to bots.  Decoy bridges are configured in
        DECOY_BRIDGES_FILE.

Changes in version 0.7.1 - 2019-06-07

        * FIXES #28496 https://bugs.torproject.org/28496
        Remove Yahoo from the list of allowed email domains.  Yahoo allows you
        to create up to 500 disposable email addresses, which BridgeDB
        interprets as unique:
        https://bugs.torproject.org/28496#comment:8
        We could address this issue in BridgeDB but at this point we seem better
        off dropping support for Yahoo because the provider likely also fell
        behind in Sybil protection.

Changes in version 0.7.0 - 2019-06-07

        * FIXES #28655 https://bugs.torproject.org/28655
        When a bridge supports an active probing-resistant transport, it should
        not give out flavors that are vulnerable to active probing.  For
        example, if a bridge supports obfs4 and obfs3, it should only give out
        obfs4.

        * FIXES #30706 https://bugs.torproject.org/30706

bridgedb/metrics.py
# -*- coding: utf-8 ; test-case-name: bridgedb.test.test_metrics ; -*-
# _____________________________________________________________________________
#
# This file is part of BridgeDB, a Tor bridge distribution system.
#
# :authors: please see included AUTHORS file
# :copyright: (c) 2019, The Tor Project, Inc.
#             (c) 2019, Philipp Winter
# :license: see LICENSE for licensing information
# _____________________________________________________________________________

"""API for keeping track of BridgeDB metrics, e.g., the demand for bridges
over time.
"""

import logging
import ipaddr
import operator
import json
import datetime

from bridgedb import geo
from bridgedb.distributors.common.http import getClientIP
from bridgedb.distributors.email import request
from bridgedb.distributors.email.distributor import EmailRequestedHelp

from twisted.mail.smtp import Address

# Our data structure to keep track of exit relays.  The variable is of type
# bridgedb.proxy.ProxySet.  We reserve a special country code (determined by
# PROXY_CC below) for exit relays and other proxies.
PROXIES = None

# Our custom country code for IP addresses that we couldn't map to a country.
# This can happen for private IP addresses or if our geo-location provider has
# no mapping.
UNKNOWN_CC = "??"

# Our custom country code for IP addresses that are proxies, e.g., Tor exit
# relays.  The code "zz" is free for assignment for user needs as specified
# here: <https://en.wikipedia.org/w/index.php?title=ISO_3166-1_alpha-2&oldid=906611218#Decoding_table>
PROXY_CC = "ZZ"

# We use BIN_SIZE to reduce the granularity of our counters.  We round up
# numbers to the next multiple of BIN_SIZE, e.g., 28 is rounded up to:
# 10 * 3 = 30.
BIN_SIZE = 10

# The prefix length that we use to keep track of the number of unique subnets
# we have seen HTTPS requests from.
SUBNET_CTR_PREFIX_LEN = 20

# All of the pluggable transports BridgeDB currently supports.
SUPPORTED_TRANSPORTS = None

# Version number for our metrics format.  We increment the version if our
# format changes.
METRICS_VERSION = 1


def setProxies(proxies):
    """Set the given proxies.

    :type proxies: :class:`~bridgedb.proxy.ProxySet`
    :param proxies: The container for the IP addresses of any currently
        known open proxies.
    """
    logging.debug("Setting %d proxies." % len(proxies))
    global PROXIES
    PROXIES = proxies


def setSupportedTransports(supportedTransports):
    """Set the given supported transports.

    :param dict supportedTransports: The transport types that BridgeDB
        currently supports.
    """

    logging.debug("Setting %d supported transports." %
                  len(supportedTransports))
    global SUPPORTED_TRANSPORTS
    SUPPORTED_TRANSPORTS = supportedTransports


def isBridgeTypeSupported(bridgeType):
    """Return `True' or `False' depending on if the given bridge type is
    supported.

    :param str bridgeType: The bridge type, e.g., "vanilla" or "obfs4".
    """

    if SUPPORTED_TRANSPORTS is None:
        logging.error("Bug: Variable SUPPORTED_TRANSPORTS is None.")
        return False

    # Note that "vanilla" isn't a transport protocol (in fact, it's the absence
    # of a transport), which is why it isn't in SUPPORTED_TRANSPORTS.
    return (bridgeType in SUPPORTED_TRANSPORTS) or (bridgeType == "vanilla")


def export(fh, measurementInterval):
    """Export metrics by writing them to the given file handle.

    :param file fh: The file handle to which we're writing our metrics.
    :param int measurementInterval: The number of seconds after which we rotate
        and dump our metrics.
    """

    httpsMetrix = HTTPSMetrics()
    emailMetrix = EmailMetrics()
    moatMetrix = MoatMetrics()

    # Rotate our metrics.
    httpsMetrix.rotate()
    emailMetrix.rotate()
    moatMetrix.rotate()

    numProxies = len(PROXIES) if PROXIES is not None else 0
    if numProxies == 0:
        logging.error("Metrics module doesn't have any proxies.")
    else:
        logging.debug("Metrics module knows about %d proxies." % numProxies)

    now = datetime.datetime.utcnow()
    fh.write("bridgedb-metrics-end %s (%d s)\n" % (
             now.strftime("%Y-%m-%d %H:%M:%S"),
             measurementInterval))
    fh.write("bridgedb-metrics-version %d\n" % METRICS_VERSION)

    httpsLines = httpsMetrix.getMetrics()
    for line in httpsLines:
        fh.write("bridgedb-metric-count %s\n" % line)

    moatLines = moatMetrix.getMetrics()
    for line in moatLines:
        fh.write("bridgedb-metric-count %s\n" % line)

    emailLines = emailMetrix.getMetrics()
    for line in emailLines:
        fh.write("bridgedb-metric-count %s\n" % line)


def resolveCountryCode(ipAddr):
    """Return the country code of the given IP address.

    :param str ipAddr: The IP address to resolve.

    :rtype: str
    :returns: A two-letter country code.
    """

    if ipAddr is None:
        logging.warning("Given IP address was None.  Using %s as country "
                        "code." % UNKNOWN_CC)
        return UNKNOWN_CC

    if PROXIES is None:
        logging.warning("Proxies are not yet set.")
    elif ipAddr in PROXIES:
        return PROXY_CC

    countryCode = geo.getCountryCode(ipaddr.IPAddress(ipAddr))

    # countryCode may be None if GeoIP is unable to map an IP address to a
    # country.
    return UNKNOWN_CC if countryCode is None else countryCode


class Singleton(type):
    _instances = {}

    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            cls._instances[cls] = super(Singleton, cls).__call__(*args,
                                                                 **kwargs)
        return cls._instances[cls]

    def clear(cls):
        """Drop the instance (necessary for unit tests)."""
        try:
            del cls._instances[cls]
        except KeyError:
            pass


class Metrics(metaclass=Singleton):
    """Base class representing metrics.

    This class provides functionality that our three distribution mechanisms
    share.
    """

    def __init__(self, binSize=BIN_SIZE):
        logging.debug("Instantiating metrics class.")
        self.binSize = binSize

        # Metrics cover a 24 hour period.  To that end, we're maintaining two
        # data structures: our "hot" metrics are currently being populated
        # while our "cold" metrics are finished, and valid for 24 hours.  After
        # that, our hot metrics turn into cold metrics, and we start over.
        self.hotMetrics = dict()
        self.coldMetrics = dict()

    def rotate(self):
        """Rotate our metrics."""

        self.coldMetrics = self.hotMetrics
        self.hotMetrics = dict()

    def findAnomaly(self, request):
        anomaly = "none"

        # TODO: Inspect email for traces of bots, Sherlock Homes-style!
        # See <https://bugs.torproject.org/9316#comment:19> for the rationale.
        # All classes that inherit from Metrics() should implement this method.

        return anomaly

    def getMetrics(self):
        """Get our sanitized current metrics, one per line.

        Metrics are of the form:

            [
             "moat.obfs4.us.success.none 10",
             "https.vanilla.de.success.none 30",
             ...
            ]

        :rtype: list
        :returns: A list of metric lines.
        """
        lines = []
        for key, value in self.coldMetrics.items():
            # Round up our value to the nearest multiple of self.binSize to
            # reduce the accuracy of our real values.
            if (value % self.binSize) > 0:
                value += self.binSize - (value % self.binSize)
            lines.append("%s %d" % (key, value))
        return lines

    def set(self, key, value):
        """Set the given key to the given value.

        :param str key: The time series key.
        :param int value: The time series value.
        """
        self.hotMetrics[key] = value

    def inc(self, key):
        """Increment the given key.

        :param str key: The time series key.
        """
        if key in self.hotMetrics:
            self.hotMetrics[key] += 1
        else:
            self.set(key, 1)

    def createKey(self, distMechanism, bridgeType, countryOrProvider,
                  success, anomaly):
        """Create and return a time series key.

        :param str distMechanism: A string representing our distribution
            mechanism, e.g., "https".
        :param str bridgeType: A string representing the requested bridge
            type, e.g., "vanilla" or "obfs4".
        :param str countryOrProvider: A string representing the client's
            two-letter country code or email provider, e.g., "it" or
            "yahoo.com".
        :param bool success: ``True`` if the request was successful and
            BridgeDB handed out a bridge; ``False`` otherwise.
        :param str anomaly: ``None`` if the request was not anomalous and hence
            believed to have come from a real user; otherwise a string
            representing the type of anomaly.
        :rtype: str
        :returns: A key that uniquely identifies the given metrics
            combinations.
        """

        if isinstance(countryOrProvider, bytes):
            countryOrProvider = countryOrProvider.decode('utf-8')

        countryOrProvider = countryOrProvider.lower()
        bridgeType = bridgeType.lower()
        success = "success" if success else "fail"

        key = "%s.%s.%s.%s.%s" % (distMechanism, bridgeType,
                                  countryOrProvider, success, anomaly)

        return key


class HTTPSMetrics(Metrics):

    def __init__(self):
        super(HTTPSMetrics, self).__init__()

        # Maps subnets (e.g., "1.2.0.0/16") to the number of times we've seen
        # requests from the given subnet.
        self.subnetCounter = dict()
        self.keyPrefix = "https"

    def getTopNSubnets(self, n=10):

        sortedByNum = sorted(self.subnetCounter.items(),
                             key=operator.itemgetter(1),
                             reverse=True)
        return sortedByNum[:n]

    def _recordHTTPSRequest(self, request, success):

        logging.debug("HTTPS request has user agent: %s" %
                      request.requestHeaders.getRawHeaders("User-Agent"))

        # Pull the client's IP address out of the request and convert it to a
        # two-letter country code.
        ipAddr = getClientIP(request,
                             useForwardedHeader=True,
                             skipLoopback=False)
        self.updateSubnetCounter(ipAddr)
        countryCode = resolveCountryCode(ipAddr)

        transports = request.args.get("transport", list())
        if len(transports) > 1:
            logging.warning("Expected a maximum of one transport but %d are "
                            "given." % len(transports))

        if len(transports) == 0:
            bridgeType = "vanilla"
        elif transports[0] == "" or transports[0] == "0":
            bridgeType = "vanilla"
        else:
            bridgeType = transports[0]

        # BridgeDB's HTTPS interface exposes transport types as a drop down
        # menu but users can still request anything by manipulating HTTP
        # parameters.
        if not isBridgeTypeSupported(bridgeType):
            logging.warning("User requested unsupported transport type %s "
                            "over HTTPS." % bridgeType)
            return

        logging.debug("Recording %svalid HTTPS request for %s from %s (%s)." %
                      ("" if success else "in",
                       bridgeType, ipAddr, countryCode))

        # Now update our metrics.
        key = self.createKey(self.keyPrefix, bridgeType, countryCode,
                             success, self.findAnomaly(request))
        self.inc(key)

    def recordValidHTTPSRequest(self, request):
        self._recordHTTPSRequest(request, True)

    def recordInvalidHTTPSRequest(self, request):
        self._recordHTTPSRequest(request, False)

    def updateSubnetCounter(self, ipAddr):

        if ipAddr is None:
            return

        nw = ipaddr.IPNetwork(ipAddr + "/" + str(SUBNET_CTR_PREFIX_LEN),
                              strict=False)
        subnet = nw.network.compressed
        logging.debug("Updating subnet counter with %s" % subnet)

        num = self.subnetCounter.get(subnet, 0)
        self.subnetCounter[subnet] = num + 1


class EmailMetrics(Metrics):

    def __init__(self):
        super(EmailMetrics, self).__init__()
        self.keyPrefix = "email"

    def _recordEmailRequest(self, smtpAutoresp, success):

        emailAddrs = smtpAutoresp.getMailTo()
        if len(emailAddrs) == 0:
            # This is just for unit tests.
            emailAddr = Address("foo@gmail.com")
        else:
            emailAddr = emailAddrs[0]

        # Get the requested transport protocol.
        try:
            br = request.determineBridgeRequestOptions(
                    smtpAutoresp.incoming.lines)
        except EmailRequestedHelp:
            return
        bridgeType = "vanilla" if not len(br.transports) else br.transports[0]

        # Over email, transports are requested by typing them.  Typos happen
        # and users can request anything, really.
        if not isBridgeTypeSupported(bridgeType):
            logging.warning("User requested unsupported transport type %s "
                            "over email." % bridgeType)
            return

        logging.debug("Recording %svalid email request for %s from %s." %
                      ("" if success else "in", bridgeType, emailAddr))
        sld = emailAddr.domain.split(b".")[0]

        # Now update our metrics.
        key = self.createKey(self.keyPrefix, bridgeType, sld, success,
                             self.findAnomaly(request))
        self.inc(key)

    def recordValidEmailRequest(self, smtpAutoresp):
        self._recordEmailRequest(smtpAutoresp, True)

    def recordInvalidEmailRequest(self, smtpAutoresp):
        self._recordEmailRequest(smtpAutoresp, False)


class MoatMetrics(Metrics):

    def __init__(self):
        super(MoatMetrics, self).__init__()
        self.keyPrefix = "moat"

    def _recordMoatRequest(self, request, success):

        logging.debug("Moat request has user agent: %s" %
                      request.requestHeaders.getRawHeaders("User-Agent"))

        ipAddr = getClientIP(request,
                             useForwardedHeader=True,
                             skipLoopback=False)
        countryCode = resolveCountryCode(ipAddr)

        try:
            encodedClientData = request.content.read()
            clientData = json.loads(encodedClientData)["data"][0]
            transport = clientData["transport"]
            bridgeType = "vanilla" if not len(transport) else transport
        except Exception as err:
            logging.warning("Could not decode request: %s" % err)
            return

        if not isBridgeTypeSupported(bridgeType):
            logging.warning("User requested unsupported transport type %s "
                            "over moat." % bridgeType)
            return

        logging.debug("Recording %svalid moat request for %s from %s (%s)." %
                      ("" if success else "in",
                       bridgeType, ipAddr, countryCode))

        # Now update our metrics.
        key = self.createKey(self.keyPrefix, bridgeType,
                             countryCode, success, self.findAnomaly(request))
        self.inc(key)

    def recordValidMoatRequest(self, request):
        self._recordMoatRequest(request, True)

    def recordInvalidMoatRequest(self, request):
        self._recordMoatRequest(request, False)
